scripts/pgs/LDpred2/fun.R

```diff
@@ -161,3 +161,47 @@ rename_columns <- function(df, old_names, new_names) {
   colnames(df)[match(old_names, colnames(df))] <- new_names
   return(df)
 }
+
+# Filter a data frame or a vector using values in a file
+# Mainly used to filter snps
+#' @param dta Either a data frame or vector
+#' @param fileFilter A file with RSIDs
+#' @param colFilter Column name in dta to filter on
+#' @param col The column name in fileSNPs if different than the first
+#' @param verbose Print progress
+#' @return Same object type as dta, filtered for entries in fileFilter
+filterFromFile <- function(dta, fileFilter, colFilter=NULL, col=NULL, verbose=T) {
+  if (is.data.frame(dta) && !(colFilter %in% colnames(dta))) stop('Missing column in data to filter: ', colFilter)
+  if (!file.exists(fileFilter)) stop('Could not find file: ', fileFilter)
+  rowsBefore <- length(dta)
+  dtaFilter <- data.table::fread(fileFilter, sep='auto', data.table=F)
+  if (is.character(col) && !(col %in% colnames(dtaFilter))) stop('Could not find column ', col, ' in file ', fileFilter)
+  if (is.integer(col) && !(col %in% 1:ncol(dtaFilter))) stop('Only columns indexed 1-', ncol(dtaFilter),' available in file ', fileFilter)
+  if (is.character(col) || is.integer(col)) dtaFilter <- dtaFilter[,col]
+  else dtaFilter <- dtaFilter[,1]
+  if (verbose) cat('Read', length(dtaFilter), 'rows from', fileFilter, '\n')
+  #if (length(dtaFilter) > 0) dta <- dta[dta %in% dtaFilter]
+  if (is.data.frame(dta)) {
+    dta <- dta[dta[,colFilter] %in% dtaFilter,]
+    nKept <- nrow(dta)
+  } else {
+    dta <- dta[dta %in% dtaFilter]
+    nKept <- length(dta)
+  }
+  if (verbose) cat('Retained', nrow(dta), 'out of', nKept, 'rows\n')
+  dta
+}
+
+# Get the betas from snp_ldpred2_auto
+#' @param fitAuto The return value form snp_ldpred2_auto
+#' @param quantile Range of estimates to keep
+getBetasAuto <- function (fitAuto, quantile=0.95, verbose=T) {
+  range <- sapply(fitAuto, function (auto) diff(range(auto$corr_est)))
+  # Keep chains that pass the filtering below
```

Comment:
@espenhgn I noted that the problem om missing values start here (l199) and not at the `quantile` function below. There needs to be a `na.rm=T` to `range` as well. This is because if there are missing values in `corr_est` there will be missing in `range` and these will inevitably lead to missing values when assigning the `keep` variable below. Right now all chains with any missingness are thrown away, and I'm unsure if that's optimal

Reply:
Not sure either. 
Don't think it matters here, but perhaps best not to define a variable `range` replacing the `base::range` function within this function.

Reply:
Yea, that's a good point, will change the name

Reply:
`na.rm=T` in `range` causes other issues if an entire chain is missing, so will leave for now and address later.